compiler/environment.ml

```diff
@@ -0,0 +1,116 @@
+open Ast;;
+
+module FunctionMap = Map.Make(String);;
+module VariableMap = Map.Make(String);;
+module ArrayTypeMap = Map.Make(String);;
+
+exception VarAlreadyDeclared;;
+exception VarNotDeclared;;
+exception FunctionAlreadyDeclared;;
+exception FunctionNotDeclared;;
+exception IncorrectFunctionParameterTypes;;
+exception MixedTypeArray;;
+exception ArrayLiteralTypeMismatch;;
+
+type func_info  = {
+  id : string; 
+  return : data_type; 
+  args : data_type list;
+  arg_names: string list;
+}
+
+type symbol_table = {
+  func_map: func_info FunctionMap.t;
+  var_map: data_type VariableMap.t;
+  array_type_map: data_type ArrayTypeMap.t;
+}
+
+let create = 
+  {
+    func_map = FunctionMap.empty;
+    var_map = VariableMap.empty;
+    array_type_map = ArrayTypeMap.empty;
+  }
+
+let update f_map v_map a_type_map =
+  {
+    func_map = f_map;
+    var_map = v_map;
+    array_type_map = a_type_map;
+  }
+
+let string_to_data_type (s : string) = match s
+  with "int" -> Int
+  | "float" -> Float
+  | "bool" -> Bool
+  | "string" -> String
+  | "array" -> Array(Int)
+  | "json" -> Json
+  | _ -> raise (Failure "unsupported data type")
+
+let declare_var (id : string) (data_type : string) (env : symbol_table) =
+  if VariableMap.mem id env.var_map then 
+    raise VarAlreadyDeclared
+  else 
+    let update_var_map = VariableMap.add id (string_to_data_type(data_type)) env.var_map in
+    update env.func_map update_var_map env.array_type_map
```

Comment:
Yeah, that seems harder, plus we're not creating a new map for each of these, we're just passing a new one. Doesn't like it's worth putting more mental attention towards.